### PR TITLE
Adds wait_for_node

### PR DESCRIPTION
**Pull request by gonzodepedro**
Related #823 

Signed-off-by: Gonzalo de Pedro <gonzalo@depedro.com.ar>

### CODE REVIEW

File: rclpy/rclpy/node.py [hunk 1]
```diff
@@ -1949,3 +1950,22 @@ def get_subscriptions_info_by_topic(
             topic_name,
             no_mangle,
             _rclpy.rclpy_get_subscriptions_info_by_topic)
+
+    def wait_for_node(
+        self,
```
Comment (jacobperron) [inline rclpy/rclpy/node.py:1955]: Another approach we could take is to put the implementation into a free function, living in it's own module (e.g. `graph.py`), that takes a node as an argument. This makes it easier to do dependency injection on node-like objects (nice for testing). We can can still have node method too, that just calls the free function.

I think this is a very minor comment that can be ignored; it's easy enough to break out the implementation later without breaking API.

**Review by jacobperron — COMMENTED** [review]
LGTM, I've left a bunch of minor comments below.

File: rclpy/rclpy/node.py [hunk 2]
```diff
@@ -1949,3 +1950,22 @@ def get_subscriptions_info_by_topic(
...
+        node_name: str,
```
Comment (fujitatomoya) [inline rclpy/rclpy/node.py:1956]: how about adding namespace as well which could be empty if user application does not use.

**Review by fujitatomoya — APPROVED** [review]

File: rclpy/rclpy/node.py [hunk 3]
```diff
@@ -1949,3 +1950,22 @@ def get_subscriptions_info_by_topic(
...
+        timeout: float = 8.0
+    ) -> bool:
+        """
+        Wait until node name is present in the system or timeout.
+
+        :param node_name: name of the node to wait for.
+        :timeout: seconds to wait for the node to be present.
```
Comment (jacobperron) [inline rclpy/rclpy/node.py:1963]: missing "param":
```suggestion
        :param timeout: seconds to wait for the node to be present.
```

File: rclpy/test/test_node.py [hunk 4]
```diff
@@ -1896,6 +1896,18 @@ def test_static_dynamic_typing(self):
         self.assertEqual(self.node.get_parameter('int_param').value, 4)
         self.assertFalse(self.node.has_parameter('dynamic_param'))
 
+    def test_wait_for_node(self):
+        try:
+            node = rclpy.create_node(
+                TEST_NODE + '_create_node', namespace='/my_ns', context=self.context
+            )
+            self.assertTrue(self.node.wait_for_node('my_node'))
```
Comment (jacobperron) [inline rclpy/test/test_node.py:1904]: We're missing the timeout argument to `wait_for_node` here.

**Review by jacobperron — COMMENTED** [review]
LGTM after the test is fixed.

File: rclpy/rclpy/node.py [hunk 5]
```diff
@@ -1949,3 +1950,25 @@ def get_subscriptions_info_by_topic(
...
+        :param timeout: seconds to wait for the node to be present. If negative, the function
+                         won't timeout.
+        :return: True if the node was found, False if timeout.
+        """
+        start = time.time()
+        flag = False
+        # TODO refactor this implementation when we can react to guard condition events, or replace
+        # it entirely with an implementation in rcl. see https://github.com/ros2/rclpy/issues/929
+        while time.time() - start < timeout and not flag:
+            flag = node_name in self.get_node_names()
```
Comment (ivanpauno) [inline rclpy/rclpy/node.py:1972]: The node name is not an unique identifier for a node, `get_node_names_and_namespaces()` should be used instead.
Having a `get_fully_qualified_node_names()` would be useful as well.

**Review by ivanpauno — CHANGES_REQUESTED** [review]

**Comment by ivanpauno** [comment]
@jacobperron are you planning to pick this up from Gonzo?
If not, I can address the remaining comments.

File: rclpy/rclpy/node.py (hunk 2 above)
Comment (jacobperron) [reply rclpy/rclpy/node.py:1956]: I've updated the documentation to clarify this should by a fully qualified name (ie. including the namespace).

File: rclpy/rclpy/node.py (hunk 5 above)
Comment (jacobperron) [reply rclpy/rclpy/node.py]: Good catch. I've updated this PR to account for fully qualified node names.

File: rclpy/rclpy/node.py [hunk 6]
```diff
@@ -1949,3 +1963,31 @@ def get_subscriptions_info_by_topic(
             topic_name,
             no_mangle,
             _rclpy.rclpy_get_subscriptions_info_by_topic)
+
+    def wait_for_node(
+        self,
+        node_name: str,
```
Comment (ivanpauno) [inline rclpy/rclpy/node.py:1969]: nit:

I would use:
```suggestion
        fq_node_name: str,
```
or `fully_qualified_node_name`, to make it more explicit.

File: rclpy/test/test_node.py [hunk 7]
```diff
@@ -1896,6 +1896,18 @@ def test_static_dynamic_typing(self):
         self.assertEqual(self.node.get_parameter('int_param').value, 4)
         self.assertFalse(self.node.has_parameter('dynamic_param'))
 
+    def test_wait_for_node(self):
+        try:
+            node = rclpy.create_node(
+                'waiting_for_this_node', namespace='/my_ns', context=self.context
+            )
+            self.assertTrue(self.node.wait_for_node('/my_ns/waiting_for_this_node', 0.5))
```
Comment (ivanpauno) [inline rclpy/test/test_node.py:1904]: I would use a larger timeout, so it works reliably in all platforms and with all rmw implementations:

```suggestion
            self.assertTrue(self.node.wait_for_node('/my_ns/waiting_for_this_node', 3))
```

**Review by ivanpauno — APPROVED** [review]
I have some minor suggestions.
LGTM!

**Comment by ivanpauno** [comment]
@jacobperron friendly ping

File: rclpy/rclpy/node.py (hunk 6 above)
Comment (jacobperron) [reply rclpy/rclpy/node.py]: Done: [90e8bf0](https://github.com/ros2/rclpy/pull/930/commits/90e8bf0f56903c40ee4a6741e85acbb5d5cedb0f)

File: rclpy/test/test_node.py (hunk 7 above)
Comment (jacobperron) [reply rclpy/test/test_node.py]: done [90e8bf0](https://github.com/ros2/rclpy/pull/930/commits/90e8bf0f56903c40ee4a6741e85acbb5d5cedb0f)

**Review by ivanpauno — APPROVED** [review]

**Comment by jacobperron** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/17110/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/11658/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/17527/)

**Pull request merged**